### DISCUSSION TITLE
Diffs for amended commit sometimes only show the new changes

### DISCUSSION THREAD

**Discussion by simon-katz** [magit/magit#4507] [Support] [answered]
I've recently updated from Magit 2.90.1 to 20210913.1931 and from Emacs 26.3 to 27.1.

Somewhere along the way I seem to have a new and strange behaviour for `magit-commit-amend`. Sometimes the diff only shows the new changes (but sometimes it shows the old + new changes).

Is this a known issue?

I'm happy to try to put together a bug report if and when I understand it better, with instructions on how to recreate the issue.

**Comment by tarsius** [comment, marked as answer]
When you initiate the commit are you sometimes in a diff buffer and sometimes not? That is the only reason I can see that might lead to different results.

While a commit is in progress you can type `C-c C-d` to show the other diff.

There are some known (and known unknown) bugs in the "show the appropriate diff while committing" logic and I intend to look into that next month or so. #3246

So there is no need to open a new issue and please use `C-c C-d` until I get around to fixing this.

**Reply by simon-katz** [reply]
I've been initiating the commit from a Magit status buffer using `c` then `a`.

Of course, I can't reproduce it at the moment. But your mention of `C-c C-d` means I know where to look next time it happens.

Thanks.

**Reply by simon-katz** [reply]
@tarsius 

I've found the problem.

I get the different behaviour when I'm amending a pushed commit.

In `magit-commit-diff`, the value of `last-command` is checked. When I'm amending a pushed commit, Magit asks me to confirm that I want to do that and the value of `last-command` is then `y-or-n-p-insert-y` rather than `magit-commit-amend`.

FWIW, `C-c C-d` doesn't seem to toggle things — rather it just sets the diff to show the staged changes.

I have hacked things locally to behave more nicely, so no hurry on any fix from my point of view.

**Reply by kyleam** [reply]
@simon-katz 
> When I'm amending a pushed commit, Magit asks me to confirm that I want to do that and the value of `last-command` is then `y-or-n-p-insert-y` rather than `magit-commit-amend`.

On my end `last-command` remains `magit-commit-amend`.  I _think_ the difference is that you're on Emacs 27.1 and I'm on 27.2.  In particular, my guess is that Emacs 27.2's a40be60ec8 (Backport Handle read-char-from-minibuffer and y-or-n-p from pre-command-hook, 2020-12-06) resolves this issue.

**Reply by simon-katz** [reply]
@kyleam You are right — with Emacs 27.2 all is good. Thanks!

Unfortunately, with Emacs 27.2 I have a worse problem: https://emacs.stackexchange.com/questions/68288/error-retrieving-https-elpa-gnu-org-packages-archive-contents

Maybe I'll look at monkey-patching the commit you mentioned into my Emacs setup for Emacs 27.1.

EDIT: I've found how to avoid that problem I had with Emacs 27.2 — see that link above for details if you are interested.
